### PR TITLE
Primary caching 3: barebone latest-at caching

### PR DESCRIPTION
**Pull request by teh-cmc**
This implements the most barebone latest-at caching support.

The goal is merely to introduce all the machinery and boilerplate required to get the primary cache running, actual caching features will be implemented on top of this foundation in follow up PRs. 

The [existing benchmark suite](https://github.com/rerun-io/rerun/blob/790f391/crates/re_query/benches/query_benchmark.rs) has been ported as-is to the cached APIs (5950X, Arch):
```
group                           primcache_3_vanilla                         primcache_3_cached                  
-----                           -------------------                         ------------------                  
arrow_batch_points2/insert      1.02  1015.0±11.07µs 939.6 MElem/sec      1.00   1000.0±7.37µs 953.7 MElem/sec
arrow_batch_points2/query       2.90      3.4±0.02µs 276.5 MElem/sec      1.00  1190.5±41.55ns 801.0 MElem/sec
arrow_batch_strings2/insert     1.00   1045.7±7.85µs 912.0 MElem/sec      1.00  1042.1±14.01µs 915.1 MElem/sec
arrow_batch_strings2/query      1.91     21.3±0.17µs  44.7 MElem/sec      1.00     11.2±0.04µs  85.2 MElem/sec 
arrow_mono_points2/insert       1.01   1789.2±3.40ms 545.8 KElem/sec      1.00  1773.6±23.00ms 550.6 KElem/sec
arrow_mono_points2/query        6.78  1102.4±18.79µs 885.9 KElem/sec      1.00    162.6±3.39µs   5.9 MElem/sec 
arrow_mono_strings2/insert      1.00   1777.3±5.89ms 549.5 KElem/sec      1.00   1777.3±7.53ms 549.5 KElem/sec
arrow_mono_strings2/query       6.30  1149.9±15.36µs 849.3 KElem/sec      1.00    182.5±0.41µs   5.2 MElem/sec 
```

---

Part of the primary caching series of PR (index search, joins, deserialization):
- #4592
- #4593
- #4659
- #4680 
- #4681
- #4698
- #4711
- #4712
- #4721 
- #4726 

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested the web demo (if applicable):
  * Using newly built examples: [app.rerun.io](https://app.rerun.io/pr/4592/index.html)
  * Using examples from latest `main` build: [app.rerun.io](https://app.rerun.io/pr/4592/index.html?manifest_url=https://app.rerun.io/version/main/examples_manifest.json)
  * Using full set of examples from `nightly` build: [app.rerun.io](https://app.rerun.io/pr/4592/index.html?manifest_url=https://app.rerun.io/version/nightly/examples_manifest.json)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/4592)
- [Docs preview](https://rerun.io/preview/bee3f9a64f1b6446e97bbcd63a650c639df0dc15/docs) 
- [Examples preview](https://rerun.io/preview/bee3f9a64f1b6446e97bbcd63a650c639df0dc15/examples) 
- [Recent benchmark results](https://build.rerun.io/graphs/crates.html)
- [Wasm size tracking](https://build.rerun.io/graphs/sizes.html)

### CODE REVIEW

File: crates/re_log_types/benches/vec_deque_ext.rs
Comment (teh-cmc) [inline crates/re_log_types/benches/vec_deque_ext.rs:1]: Just a rename of `VecDeque::insert_range` to `VecDeque::insert_many` to match the terminology in `FlatVecDeque`.

File: crates/re_log_types/src/vec_deque_ext.rs
Comment (teh-cmc) [inline crates/re_log_types/src/vec_deque_ext.rs:1]: Just a rename of `VecDeque::insert_range` to `VecDeque::insert_many` to match the terminology in `FlatVecDeque`.

File: crates/re_query_cache/benches/latest_at.rs
Comment (teh-cmc) [inline crates/re_query_cache/benches/latest_at.rs:1]: This is a literal port of [our existing benchmarks](https://github.com/rerun-io/rerun/blob/790f391/crates/re_query/benches/query_benchmark.rs), only using the cached query APIs instead.

File: crates/re_query_cache/src/query.rs [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+// ---
+
+/// Either a reference to a slice of data residing in the cache, or some data being deserialized
+/// just-in-time from an [`re_query::ArchetypeView`].
+#[derive(Debug, Clone)]
+pub enum MaybeCachedComponentData<'a, C> {
+    Cached(&'a [C]),
+    // TODO(cmc): Ideally, this would be a reference to a `dyn Iterator` that is the result of
+    // calling `ArchetypeView::iter_{required|optional}_component`.
+    // In practice this enters lifetime invariance hell for, from what I can see, no particular gains.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/query.rs:21]:
```suggestion
    // In practice this enters lifetime invariance hell for, from what I can see, no particular gains
    // (rustc is pretty good at optimizing out collections into obvious temporary variables).

File: crates/re_query_cache/src/query.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    Raw(Vec<C>),
+}
+
+impl<'a, C: Clone> MaybeCachedComponentData<'a, C> {
+    #[inline]
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = &C> + '_ {
```
Comment (teh-cmc) [inline crates/re_query_cache/src/query.rs:27]: Do we still need this? If so, just write it in terms of `Self::as_slice`

File: crates/re_query_cache/src/query.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+            );
+
+            match &query {
+                AnyQuery::LatestAt(query) => {
+                    Caches::with_latest_at::<A, _, _>(
+                        store.id().clone(),
+                        entity_path.clone(),
+                        query,
+                        |cache| {
+                            re_tracing::profile_scope!("latest_at");
```
Comment (teh-cmc) [inline crates/re_query_cache/src/query.rs:107]: can we _easily_ get this out without `rustc` going nuts to fight the righwards madness?

File: crates/re_query_cache/tests/latest_at.rs
Comment (teh-cmc) [inline crates/re_query_cache/tests/latest_at.rs:1]: This is a literal port of [our existing test suite](https://github.com/rerun-io/rerun/blob/790f391/crates/re_query/tests/archetype_query_tests.rs), only using the cached query APIs instead.

File: crates/re_query_cache/src/cache.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+// ---
+
+// TODO(cmc): Centralize and harmonize all caches (query, jpeg, mesh).
+static CACHES: Lazy<Caches> = Lazy::new(Caches::default);
+
+/// Maintains the top-level cache mappings.
+//
+// TODO(cmc): Store subscriber and cache invalidation.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:41]: - Implemented in https://github.com/rerun-io/rerun/pull/4726

File: crates/re_query_cache/src/cache.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+// TODO(cmc): SizeBytes support + size stats + mem panel
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:42]: Open and link issue

File: crates/re_query_cache/src/cache.rs [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+// TODO(cmc): timeless caching support
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:43]: - Implemented in https://github.com/rerun-io/rerun/pull/4721

File: crates/re_query_cache/src/cache.rs [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+#[derive(Default)]
+pub struct Caches {
+    latest_at: RwLock<HashMap<CacheKey, Arc<RwLock<LatestAtCache>>>>,
+}
+
+impl Caches {
+    /// Clears all caches.
+    //
+    // TODO(cmc): expose palette command.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:52]: Open and link issue

File: crates/re_query_cache/src/cache.rs [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+    ///
+    /// Reminder: within a single timestamp, rows are sorted according to their [`RowId`]s.
+    pub(crate) pov_times: VecDeque<(TimeInt, RowId)>,
+
+    /// The [`InstanceKey`]s of the point-of-view components.
+    pub(crate) pov_instance_keys: FlatVecDeque<InstanceKey>,
+
+    /// The resulting component data, pre-deserialized, pre-joined.
+    //
+    // TODO(cmc): Don't denormalize auto-generated instance keys.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:153]: Open and link issue

File: crates/re_query_cache/src/cache.rs [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    // TODO(cmc): Don't denormalize splatted values.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:154]: Open and link issue

File: crates/re_query_cache/src/cache.rs [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+    pub(crate) components: BTreeMap<ComponentName, Box<dyn ErasedFlatVecDeque + Send + Sync>>,
+    //
+    // TODO(cmc): secondary cache
+    // TODO(cmc): size stats: this requires codegen'ing SizeBytes for all components!
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:158]: Open and link issue

File: crates/re_query_cache/src/cache.rs [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+        let Some(data) = data else {
+            return Ok(Either::Left(std::iter::empty()));
+        };
+
+        Ok(Either::Right(data.iter()))
+    }
+
+    /// How many timestamps' worth of data is stored in this bucket?
+    #[inline]
+    pub fn len(&self) -> usize {
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:212]: `len` is a horrible name

File: crates/re_query_cache/src/cache.rs [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+
+/// Caches the results of `LatestAt` queries.
+///
+/// The `TimeInt` in the index corresponds to the timestamp of the query, _not_ the timestamp of
+/// the resulting data!
+//
+// TODO(cmc): we need an extra indirection layer so that cached entries can be shared across
+// queries with different query timestamps but identical data timestamps.
+// This requires keeping track of all `RowId`s in `ArchetypeView`, not just the `RowId` of the
+// point-of-view component.
```
Comment (teh-cmc) [inline crates/re_query_cache/src/cache.rs:346]: - Implemented in https://github.com/rerun-io/rerun/pull/4712

File: crates/re_query_cache/src/query.rs [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+                            if bucket.is_empty() {
+                                let now = web_time::Instant::now();
+                                let arch_view = query_archetype::<A>(store, &query, entity_path)?;
+
+                                bucket.[<insert_pov $N _comp$M>]::<A, $($pov,)+ $($comp,)*>(query.at, &arch_view)?;
+
+                                // TODO(cmc): I'd love a way of putting this information into
+                                // the `puffin` span directly.
+                                let elapsed = now.elapsed();
+                                ::re_log::trace!(
```
Comment (teh-cmc) [inline crates/re_query_cache/src/query.rs:122]: for what key?!

File: crates/re_query_cache/benches/latest_at.rs [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+    batch_strings
+);
+criterion_main!(benches);
+
+// ---
+
+fn mono_points(c: &mut Criterion) {
+    // Each mono point gets logged at a different path
+    let paths = (0..NUM_POINTS)
+        .map(move |point_idx| entity_path!("points", point_idx.to_string()))
```
Comment (emilk) [inline crates/re_query_cache/benches/latest_at.rs:57]:
```suggestion
        .map(move |point_idx| entity_path!("points", point_idx))
```

should work

File: crates/re_query_cache/src/cache.rs [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+        Self::LatestAt(query)
+    }
+}
+
+// ---
+
+// TODO(cmc): Centralize and harmonize all caches (query, jpeg, mesh).
+static CACHES: Lazy<Caches> = Lazy::new(Caches::default);
+
+/// Maintains the top-level cache mappings.
```
Comment (emilk) [inline crates/re_query_cache/src/cache.rs:39]: Are we expected to have one cache per `DataStore`?

File: crates/re_query_cache/src/cache.rs [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+    #[inline]
+    pub fn iter_pov_instance_keys(&self) -> impl Iterator<Item = &[InstanceKey]> {
+        self.pov_instance_keys.iter()
+    }
+
+    /// Iterate over the batches of the specified non-optional component.
+    #[inline]
+    pub fn iter_component<C: Component + Send + Sync + 'static>(
+        &self,
+    ) -> re_query::Result<impl Iterator<Item = &[C]>> {
```
Comment (emilk) [inline crates/re_query_cache/src/cache.rs:178]: We could consider using a `Result<_, ComponentNotFoundError>` instead, with an `impl From<ComponentNotFoundError> for QueryError` to make it more clear that the functions in this file only have one faillure care

File: crates/re_query_cache/src/cache.rs [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+            .components
+            .get(&C::name())
+            .map(|data| data.as_any().downcast_ref::<FlatVecDeque<C>>())
+            .ok_or_else(|| re_query::QueryError::RequiredComponentNotFound(C::name()))?;
+
+        let Some(data) = data else {
+            return Ok(Either::Left(std::iter::empty()));
+        };
+
+        Ok(Either::Right(data.iter()))
```
Comment (emilk) [inline crates/re_query_cache/src/cache.rs:189]: nit/style: I find this easier to read

```suggestion
        if let Some(data) = data {
            Ok(Either::Right(data.iter()))
        } else {
            Ok(Either::Left(std::iter::empty()))
        };
```

File: crates/re_query_cache/src/cache.rs [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+    /// Which [`EntityPath`] is the query targeting?
+    pub entity_path: EntityPath,
+
+    /// Which [`Timeline`] is the query targeting?
+    pub timeline: Timeline,
+
+    /// Which [`Archetype`] are we querying for?
+    ///
+    /// This is very important because of our data model: we not only query for components, but we
+    /// query for components from a specific point-of-view (the so-called primary component).
```
Comment (emilk) [inline crates/re_query_cache/src/cache.rs:101]: We should probably pick one terminology and stick with it ("primary" or "pov")

File: crates/re_query_cache/src/query.rs [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+impl<'a, C: Clone> MaybeCachedComponentData<'a, C> {
+    #[inline]
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = &C> + '_ {
+        self.as_slice().iter()
+    }
+
+    #[inline]
+    pub fn as_slice(&self) -> &[C] {
```
Comment (emilk) [inline crates/re_query_cache/src/query.rs:33]: We should also implement `Deref for MaybeCachedComponentData` with `type Target = [C];`

File: crates/re_query_cache/src/query.rs [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+                let bucket = cache.entry(query.at).or_default();
+                // NOTE: Implicitly dropping the write guard here: the LatestAtCache is free once again!
+
+                if bucket.is_empty() {
+                    let now = web_time::Instant::now();
+                    let arch_view = query_archetype::<A>(store, &query, entity_path)?;
+
+                    bucket.[<insert_pov $N _comp$M>]::<A, $($pov,)+ $($comp,)*>(query.at, &arch_view)?;
+
+                    // TODO(cmc): I'd love a way of putting this information into the `puffin` span directly.
```
Comment (emilk) [inline crates/re_query_cache/src/query.rs:110]: You can, with `re_tracing::profile_scope!("latest_at", format!(…));`

**Review by emilk — APPROVED** [review]

File: crates/re_query_cache/src/cache.rs (hunk 15 above)
Comment (teh-cmc) [reply crates/re_query_cache/src/cache.rs:39]: `CACHES` cover everything, the `StoreId` is part of the `CacheKey`.

File: crates/re_query_cache/src/cache.rs (hunk 18 above)
Comment (teh-cmc) [reply crates/re_query_cache/src/cache.rs:101]: We've discussed this during standpoint: we'll continue down that road for now and see where it leads... :grimacing:

File: crates/re_query_cache/src/cache.rs (hunk 15 above)
Comment (emilk) [reply crates/re_query_cache/src/cache.rs:39]: Put that in the docstring :)

File: crates/re_query_cache/src/cache.rs (hunk 16 above)
Comment (teh-cmc) [reply crates/re_query_cache/src/cache.rs:178]: I guess if the only possible error is NotFound I might as well return an Option instead

File: crates/re_query_cache/src/cache.rs (hunk 15 above)
Comment (teh-cmc) [reply crates/re_query_cache/src/cache.rs:38]: Done in next PR

**Pull request merged**